### PR TITLE
Code cleanup

### PR DESCRIPTION
**Pull request by oleg-derevenetz**
* Remove unused code
* Get the virtual methods in order
* Make methods const if possible
* Add missing copy constructors and assignment operators (mostly =deleted), as well as destructors (=default)
* Use =default constructors and destructors instead of empty ({})
* Use explicit keyword to disable converting constructors where possible
* Make function parameters reference-to-const where possible

### CODE REVIEW

**Review by ihhub — APPROVED** [review]

**Comment by ihhub** [comment]
Hi @oleg-derevenetz , I approved this pull request but in the future please separate changes into few pull requests as checking such big PR is troublesome. Also removing `virtual` word is definitely subjective as some want to keep it, other don't but I won't argue here.

@oleg-derevenetz , could you please review this pull request as many parts of your code were touched?

**Comment by oleg-derevenetz** [comment]
Hi @ihhub

> I approved this pull request but in the future please separate changes into few pull requests as checking such big PR is troublesome.

Oh, sorry :( This is mostly just a result of automated scripts (except the first commit) that I made to fix multiple issues pointed out by SonarCloud. I just applied it to all sources at once (except 3rd party code). Pure mechanical job :) In the future, I will try to break such changes to a separate PRs based on folder maybe(?).

> Also removing virtual word is definitely subjective as some want to keep it, other don't but I won't argue here.

I think there is no point in having both virtual and override, it's just a waste of screen (and disk) space :) `override` implies virtual and guarantees that there is a corresponding virtual method somewhere up the hierarchy. `virtual` guarantees literally nothing per se, so I think that it should be used with caution only in those rare cases when it's the first virtual method in hierarchy. It's the same as with copy/move constructors and assignment operators - I prefer to `=delete` them by default in all classes without an exception (and explicitly define them, maybe as =default if applicable, only if it is absolutely necessary) to ensure that there is no accidental copying somewhere.

**Comment by ihhub** [comment]
You're right. I was typing from my phone...
@idshibanov , could you please review this pull request as some parts of your code were changed?

**Comment by ihhub** [comment]
Hi @oleg-derevenetz , could you please rebase your branch against master? I have compilation errors with your build.

File: src/fheroes2/spell/spell.h [hunk 1]
```diff
@@ -114,7 +114,6 @@ class Spell
 
     Spell( int = NONE );
 
-    bool operator<( const Spell & ) const;
```
Comment (ihhub) [inline src/fheroes2/spell/spell.h:117]: This change causes compilation error.

**Review by ihhub — CHANGES_REQUESTED** [review]

**Comment by oleg-derevenetz** [comment]
Hi @ihhub

> This change causes compilation error.

Yes, now this operator is used by recently merged #3203. Resurrected it back.

**Comment by ihhub** [comment]
Hi @oleg-derevenetz , I had a chat with @idshibanov and he pointed out that some deleted parts of the code are needed for the future development. I would like to move forward with this pull request so if it won't be too troublesome for you could you please create a pull requets for all changes except unused code removal? It could be a separate PR. This pull request could contain only unused code removal so it'll be easier for everyone.

**Comment by oleg-derevenetz** [comment]
Hi @ihhub OK, I'll break this PR in two pieces. However, is it possible to point out later what parts of code that deleted in this PR should be preserved? There is too much of unused trivial code that is obviously obsolete, for example, various operator<</>>() from battle-related code which are no longer load/save objects properly because the set of object fields has changed long ago while those operators are not and so on. I didn't touch non-trivial code related to map/tile processing BTW.

**Comment by oleg-derevenetz** [comment]
Hi @ihhub moved commits from this PR to #3227. I'll rebase this one once #3227 will be merged.

**Comment by oleg-derevenetz** [comment]
Rebased this.

**Comment by ihhub** [comment]
> Hi @ihhub OK, I'll break this PR in two pieces. However, is it possible to point out later what parts of code that deleted in this PR should be preserved? There is too much of unused trivial code that is obviously obsolete, for example, various operator<</>>() from battle-related code which are no longer load/save objects properly because the set of object fields has changed long ago while those operators are not, constructors that are never called with such arguments and so on. I didn't touch non-trivial code related to map/tile processing BTW.

Hi @oleg-derevenetz , we would like to preserve battle related code. To move this pull request further could you please create a new one without touching `world_pathfinding.*`, `world_regions.*`, `army*` and `battle*` files? These files we can keep here for now. We'll need in the future to send commands over the network so we can't simply delete these functions for now. My apologies for such multi-step approach :(

**Comment by oleg-derevenetz** [comment]
Hi @ihhub

> To move this pull request further could you please create a new one without touching world_pathfinding.*, world_regions.*, army* and battle* files?

Created #3275.

> We'll need in the future to send commands over the network so we can't simply delete these functions for now.

There is a lot of trivial unused code in those files unrelated to future network interaction, for example, `Troops::GetAttack()`, `Troops::GetDefense()`, `Troops::GetDamageMin()`, `Troops::GetDamageMax()` and so on. But let's do a cleanup piece by piece :)

> My apologies for such multi-step approach :(

No problem :) I'll rebase this after merging of #3275.

**Comment by ihhub** [comment]
Hi @oleg-derevenetz , I checked all functions and this is a list of them which we can surely remove:
- `KeepOnlyWeakest` and `KeepOnlyWeakestTroops`
- `ArmyStrongerThanEnemy`
- `Battle::Arena::NetworkTurn` and `Battle::Interface::NetworkTurn`
- `GetPassableQualityPositions`
- `InvertedFade`
- constructors and assignment operators

**Comment by oleg-derevenetz** [comment]
Hi @ihhub Thank you! I'll make a separate PR today then.

**Comment by ihhub** [comment]
@oleg-derevenetz , thanks a lot for this code cleanup! I'm giving green light as we could still revert this code if needed based on the commit changes.

**Pull request merged**